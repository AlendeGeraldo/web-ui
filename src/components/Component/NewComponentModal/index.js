import React from 'react';
import PropTypes from 'prop-types';
import { Modal, Form, Input } from 'antd';

const NewComponentModal = (props) => {
  const { visible, form, history } = props;
  const { onCreate, onCancel } = props;
  const { getFieldDecorator, getFieldsError } = form;

  /* 
  This function is used to check if form has errors
*/
  const hasErrors = (fieldsError) => {
    return Object.keys(fieldsError).some((field) => fieldsError[field]);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    form.validateFields(async (err, values) => {
      if (err) {
        return;
      }
      onCreate(values.name, history);
    });
  };

  return (
    <Modal
      visible={visible}
      title='Novo Componente'
      okText='Criar'
      cancelText='Cancelar'
      onCancel={onCancel}
      onOk={handleSubmit}
      okButtonProps={{ disabled: hasErrors(getFieldsError()) }}
    >
      <Form layout='vertical'>
        <Form.Item label='Qual o nome do seu componente?'>
          {getFieldDecorator('name', {
            rules: [
              {
                required: true,
                message: 'Por favor insira um nome para o componente!',
              },
            ],
          })(<Input allowClear autoFocus />)}
        </Form.Item>
      </Form>
    </Modal>
  );
};

NewComponentModal.propTypes = {
  visible: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  onCreate: PropTypes.func.isRequired,
  form: PropTypes.shape({
    getFieldDecorator: PropTypes.func,
    getFieldsError: PropTypes.func,
  }).isRequired,
};

export default Form.create({ name: 'new-component' })(NewComponentModal);
