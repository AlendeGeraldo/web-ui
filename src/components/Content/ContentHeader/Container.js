// CORE LIBS
import React from 'react';
import { useHistory } from 'react-router-dom';

// COMPONENTS
import ContentHeader from './index';

/**
 * Content Header Container.
 * This component is responsible for create a logic container for content header
 * with route control.
 */
const ContentHeaderContainer = ({ title, editable }) => {
  // getting history
  const history = useHistory();

  // HANDLERS
  const goBackHandler = () => history.goBack();

  // RENDER
  return (
    <ContentHeader
      title={title}
      editable={editable}
      handleGoBack={goBackHandler}
    />
  );
};

// EXPORT
export default ContentHeaderContainer;
