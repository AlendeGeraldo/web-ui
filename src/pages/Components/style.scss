.componentsPage {
  height: 100%;
  width: 100%;
}

.componentsPageBody {
  margin: 40px;
  padding: 20px;
  background-color: #fff;
  .header {
    margin-bottom: 20px;
  }
}
