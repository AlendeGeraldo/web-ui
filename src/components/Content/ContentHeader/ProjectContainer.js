// CORE LIBS
import React, { useLayoutEffect } from 'react';
import { connect } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';

// COMPONENTS
import ContentHeader from './index';

// ACTIONS
import { fetchProject, editProjectName } from '../../../store/project/actions';

// DISPATCHS
const mapDispatchToProps = (dispatch) => {
  return {
    handleFetchProject: (projectUuid) => dispatch(fetchProject(projectUuid)),
    handleEditProjectName: (newName) =>
      dispatch(editProjectName(null, newName)),
  };
};

// STATES
const mapStateToProps = (state) => {
  return { project: state.project };
};

/**
 * Content Header Project Container.
 * This component is responsible for create a logic container for project content
 * header with route control.
 */
const ContentHeaderProjectContainer = ({
  project,
  handleFetchProject,
  handleEditProjectName,
}) => {
  // CONSTANTS
  // getting history
  const history = useHistory();
  // getting project uuid
  const { projectUuid } = useParams();

  console.log(projectUuid);

  // HOOKS
  // did mount hook
  useLayoutEffect(() => {
    // fetching projects
    handleFetchProject(projectUuid);
  }, []);

  // HANDLERS
  // go back
  const goBackHandler = () => history.goBack();

  // RENDER
  return (
    <ContentHeader
      title={project.name}
      editable
      handleGoBack={goBackHandler}
      handleSubmit={handleEditProjectName}
    />
  );
};

// EXPORT
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ContentHeaderProjectContainer);
