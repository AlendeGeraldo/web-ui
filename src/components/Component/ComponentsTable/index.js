/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';

import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Popconfirm, Table } from 'antd';

import './style.scss';

// eslint-disable-next-line
class ComponentTable extends React.Component {
  render() {
    const { componentList, onDelete } = this.props;

    const tableColumns = [
      {
        title: 'Nome do Componente',
        dataIndex: 'name',
        key: 'name',
        render: (value, record) => (
          <Link to={`components/${record.uuid}`}>{value}</Link>
        ),
        sorter: (a, b) => a.name.localeCompare(b.name),
        ellipsis: true,
      },
      {
        title: 'Data de Criação',
        dataIndex: 'createdAt',
        key: 'createdAt',
        width: 200,
        render: (value) => new Date(value).toLocaleString(),
        sorter: (a, b) => new Date(a.createdAt) - new Date(b.createdAt),
        ellipsis: true,
      },
      {
        title: '',
        key: 'action',
        width: 100,
        render: (record) => (
          <Popconfirm
            value={record}
            title={`Tem certeza de que deseja excluir o componente ${record.name} ?`}
            onConfirm={() => {
              onDelete(record);
            }}
            okText='Sim'
            cancelText='Não'
            placement='left'
          >
            <a href='#'>Remover</a>
          </Popconfirm>
        ),
      },
    ];

    return (
      <Table
        className='componentsTable'
        rowKey={(record) => record.uuid}
        dataSource={componentList}
        columns={tableColumns}
        pagination={{ pageSize: 9 }}
        scroll={{ y: 'calc(100vh - 480px)' }}
      />
    );
  }
}

ComponentTable.propTypes = {
  componentList: PropTypes.arrayOf(PropTypes.object).isRequired,
  onDelete: PropTypes.func.isRequired,
};

export default ComponentTable;
