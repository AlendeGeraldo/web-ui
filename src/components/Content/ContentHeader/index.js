// CORE LIBS
import React from 'react';
import PropTypes from 'prop-types';

// UI LIBS
import { PageHeader } from 'antd';

// COMPONENTS
import EditableTitle from '../EditableTitle';

// STYLES
import './style.scss';

// editable title class name const
const editableTitleClassName =
  'ant-page-header-heading-title autosize-input-custom';

/**
 * Content Header.
 * This component is responsible for displaying the content header with editable
 * title and go back arrow.
 */
const ContentHeader = ({ handleGoBack, editable, title }) => {
  // COMPONENTS RENDERS
  // common title
  const renderTitle = () => title;
  // editable title
  const renderEditableTitle = () => (
    // editable title component
    <EditableTitle
      loading={false}
      title={title}
      handleSubmit={(titleSubmited) => alert(titleSubmited)}
      className={editableTitleClassName}
      editingClassName={`${editableTitleClassName} edit-mode`}
    />
  );

  // RENDER
  return (
    // page header component
    <PageHeader
      className='contentHeader'
      title={editable ? renderEditableTitle() : renderTitle()}
      onBack={handleGoBack}
    />
  );
};

// PROP TYPES
ContentHeader.propTypes = {
  /** content header go back function */
  handleGoBack: PropTypes.func.isRequired,
  /** content header has editable title */
  editable: PropTypes.bool.isRequired,
  /** content header title */
  title: PropTypes.string.isRequired,
};

// EXPORT
export default ContentHeader;
