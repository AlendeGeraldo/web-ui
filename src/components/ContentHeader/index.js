import './style.scss';
import React from 'react';
import { PageHeader } from 'antd';
// import PropTypes from 'prop-types';

const ContentHeader = ({
  title,
  onTitleDoubleClick,
  subTitle,
  onBack,
  breadcrumbs,
}) => {
  const componentTitle = onTitleDoubleClick ? (
    <span onDoubleClick={onTitleDoubleClick}>{title}</span>
  ) : (
    title
  );

  const routes = breadcrumbs;

  return (
    <PageHeader
      className='contentHeader'
      title={componentTitle}
      onBack={onBack}
      breadcrumb={{ routes }}
    />
  );
};

// ContentHeader.propTypes = {
//   title: PropTypes.string,
//   subTitle: PropTypes.string,
// };

// ContentHeader.defaultProps = {
//   title: '',
//   subTitle: '',
// };

export default ContentHeader;
