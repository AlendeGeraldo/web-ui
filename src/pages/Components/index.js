import './style.scss';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Button, Empty, Spin } from 'antd';
import ComponentsTable from '../../components/Component/ComponentsTable';
import ContentHeader from '../../components/ContentHeader';
import NewComponentModal from '../../components/Component/NewComponentModal';
import emptyPlaceholder from '../../assets/emptyPlaceholder.png';
import {
  addComponent,
  deleteComponent,
  fetchComponents,
  toggleModal,
} from '../../store/actions/componentsActions';

const Components = (props) => {
  const { componentList, loading, modalIsVisible, history } = props;
  const {
    onAddComponent,
    onDeleteComponent,
    onFetchComponents,
    onToggleModal,
  } = props;

  useEffect(() => {
    onFetchComponents();
  }, []);

  const handleDelete = async (component) => {
    onDeleteComponent(component.uuid);
  };

  const renderBody = () => {
    if (loading) return <Spin />;

    return componentList.length === 0 ? (
      <Empty
        image={emptyPlaceholder}
        imageStyle={{
          height: 136,
        }}
        description={
          <span>
            <span>
              <strong>Nenhum componente foi criado</strong>
            </span>
            <br />
            <span>
              Clique no botão &quot;Novo Componente&quot; para começar
            </span>
          </span>
        }
      />
    ) : (
      <ComponentsTable componentList={componentList} onDelete={handleDelete} />
    );
  };

  const handleCreate = (name) => {
    onAddComponent(name, history);
  };

  return (
    <div className='componentsPage'>
      <NewComponentModal
        visible={modalIsVisible}
        onCreate={handleCreate}
        onCancel={onToggleModal}
      />

      <ContentHeader title='Componentes' />

      <div className='componentsPageBody'>
        <div className='header'>
          <Button
            disabled={loading}
            onClick={onToggleModal}
            type='primary'
            icon='plus'
          >
            Novo Compomente
          </Button>
        </div>
        <div className='body'>{renderBody()}</div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    ...state.components,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onAddComponent: (post, history) => {
      dispatch(addComponent(post, history));
    },
    onDeleteComponent: (id) => {
      dispatch(deleteComponent(id));
    },
    onFetchComponents: () => {
      dispatch(fetchComponents());
    },
    onToggleModal: () => {
      dispatch(toggleModal());
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Components);
