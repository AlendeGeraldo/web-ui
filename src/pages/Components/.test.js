import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import { mount, shallow } from 'enzyme';

import { act } from 'react-dom/test-utils';

import { Button, Empty, Spin } from 'antd';

import NewComponentModal from '../../components/Component/NewComponentModal';
import ComponentsTable from '../../components/Component/ComponentsTable';
import ContentHeader from '../../components/ContentHeader';
import Components from '.';

describe('Page Components should', () => {
  it('is expected render without crashing', () => {
    shallow(<Components />);
  });

  it('is expected render html correctly', () => {
    const wrapper = shallow(<Components />);

    expect(wrapper).toMatchSnapshot();
  });

  it('is expected to have a NewComponentModal child', () => {
    const wrapper = shallow(<Components />);

    expect(
      wrapper
        .children(NewComponentModal)
        .exists()
    ).toBeTruthy();
  });

  it('is expected to have a ContentHeader child', () => {
    const wrapper = shallow(<Components />);

    expect(
      wrapper
        .children(ContentHeader)
        .exists()
    ).toBeTruthy();
  });

  it('is expected to have a div child', () => {
    const wrapper = shallow(<Components />);

    expect(wrapper.children('div').exists()).toBeTruthy();
  });

  it('componentsPageBody is expected to have header', () => {
    const wrapper = shallow(<Components />);

    expect(
      wrapper
      .children('div')
      .find('.header')
      .exists()
    ).toBeTruthy();
  });

  it('componentsPageBody header is expected to have button', () => {
    const wrapper = shallow(<Components />);

    expect(
      wrapper
      .children('div')
      .find('.header')
      .children(Button)
      .exists()
    ).toBeTruthy();
  });

  it('componentsPageBody is expected to have body', () => {
    const wrapper = shallow(<Components />)

    expect(
      wrapper
      .children('div')
      .find('.body')
      .exists()
    ).toBeTruthy();
  });

  it('is expected to exist Empty', () => {
    act(() => {
      const wrapper = shallow(<Components />);
      wrapper.setState({ loading: false, componentList: []});
      expect(
        wrapper
        .find(Empty)
        .exists()
      ).toBeTruthy();
    });
  });

  it('is expected to exist ComponentsTable', () => {
    const components = [
      {
        uuid: '9014c0e6-534b-4e7d-a8db-cd5c9d8ee540',
        createdAt: '2019-11-04T17:16:53.000Z',
        updatedAt: null,
        name: 'teste 2',
        file: '',
      },
    ];

    act(() => {
      const wrapper = shallow(<Components />);
      wrapper.setState({ loading: false, componentList: components});
      expect(
        wrapper
        .find(ComponentsTable)
        .exists()
      ).toBeTruthy();
    });
  });
});
