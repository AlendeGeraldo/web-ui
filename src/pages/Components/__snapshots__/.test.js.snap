// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Components should is expected render html correctly 1`] = `
<div
  className="componentsPage"
>
  <Form(NewComponentModal)
    onCancel={[Function]}
    onCreate={[Function]}
    visible={false}
    wrappedComponentRef={[Function]}
  />
  <ContentHeader
    title="Componentes"
  />
  <div
    className="componentsPageBody"
  >
    <div
      className="header"
    >
      <Button
        block={false}
        disabled={true}
        ghost={false}
        htmlType="button"
        icon="plus"
        loading={false}
        onClick={[Function]}
        type="primary"
      >
        Novo Compomente
      </Button>
    </div>
    <div
      className="body"
    >
      <Spin
        size="default"
        spinning={true}
        wrapperClassName=""
      />
    </div>
  </div>
</div>
`;
